Arbitrary order bases on Pyramids [pyramid-dev]

mlstowell opened the pull request:
This PR adds arbitrary order H1, ND, RT, and L2 basis functions on pyramids. The implementation is based on the 2015 paper "Orientation embedded high order shape functions for the exact sequence elements of all shapes" by Fuentes et al. This paper relies on multiple levels of nested functions and, for this reason, can be difficult to follow. The implementation attempts to adhere to this nested structure wherever possible so that the paper may be used as a guide to better understanding the code.

<!--GHEX{"author":"mlstowell","assignment":"2024-11-13T01:17:01","editor":"tzanio","id":4460,"reviewers":["hughcars","tzanio","psocratis"],"merge":"2024-12-04T01:17:01","approval":"2025-03-26T02:06:08"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4460](https://github.com/mfem/mfem/pull/4460) | @mlstowell | @tzanio | @hughcars + @tzanio + @psocratis | 11/12/24 | 3/25/25 | ⌛due 12/3/24 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

brendankeith commented:
Bravo, @mlstowell!

I'm looking forward to reviewing this. :)

mdavids-cfs commented:
Hi @mlstowell , just out of curiosity: is this feature mostly usable at the moment? I think it's a really really cool and important addition. For example, in our cases, it's really difficult to get around mixed meshes where in turn it's difficult to get around pyramids to link HEX and TET portions of the mesh. Would love to give this dev branch a try for our EM modeling problems...
Mathias

mlstowell commented:
Hi, @mdavids-cfs,
It is usable but not fully functional. The main missing feature that comes to mind is support for non-conforming mesh refinement. We can have pyramids in a non-conforming mesh but we cannot refine the pyramid elements themselves, if I remember correctly.
The other issue worth mentioning is that pyramids can be temperamental. They seem to converge at the proper rates but the interpolations can look poor near the apex of pyramids. This is particularly true for distorted pyramids. Pyramids with equilateral triangular faces seem to perform best.
If you're willing to give them a try with these caveats I'd be very interested in your feedback.
Best wishes,
Mark

mdavids-cfs commented:
Awesome, thanks for the clarification. I will definitely give this a try and share some results on performance here!

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

tzanio commented:
@brendankeith and @koning -- this is ready for review 🎉

brendankeith commented:
> @brendankeith and @koning -- this is ready for review 🎉

I will get to this over the Xmas holidays 🎅

tzanio commented:
@koning, can you confirm if this works for you?

tzanio commented:
@brendankeith -- friendly reminder :)

brendankeith commented:
> @brendankeith -- friendly reminder 😄

Thanks for the reminder, @tzanio. I'm finishing exam grading today. Xmas holidays start tomorrow 😉

brendankeith commented on fem/fe/fe_h1.hpp:
```diff
@@ -148,6 +149,151 @@ class H1_WedgeElement : public NodalFiniteElement
+   mutable DenseMatrix du, ddu;
+#endif
+   DenseMatrixInverse Ti;
+   /*
```
What is the plan for the blocks of commented code here?

mlstowell replied:
No longer needed. I have removed it. Thanks!

brendankeith commented on fem/intrules.cpp:
```diff
@@ -1794,7 +1794,7 @@ IntegrationRule *IntegrationRules::PyramidIntegrationRule(int Order)
    // rule for a cube which seems to be adequate for now. When we
    // implement high order finite elements for pyramids we should
```
The comment should be updated since MFEM now has high-order elements for pyramids. I wish I knew a better integration rule to suggest

tzanio commented:
@brendankeith, @koning  and @psocratis -- Reminder to please review this when you get a chance. We would like to merge it soon. Thanks in advance for your help!

brendankeith commented:
> @brendankeith, @koning and @psocratis -- Reminder to please review this when you get a chance. We would like to merge it soon. Thanks in advance for your help!

You got it, @tzanio!

brendankeith commented on fem/eltrans.cpp:
```diff
@@ -444,6 +444,8 @@ int IsoparametricTransformation::OrderJ() const
       case FunctionSpace::Qk:
          return (FElem->GetOrder());
+      case FunctionSpace::Uk:
+         return (FElem->GetOrder()); // Experimental
```
The "experimental" comment may no longer be appropriate here and for other cases of `FunctionSpace::Uk` below

brendankeith commented on fem/fe/fe_fixed_order.hpp:
```diff
@@ -1018,6 +1018,34 @@ class Nedelec1PyrFiniteElement : public VectorFiniteElement
 
 
+/// A 3D 2nd order Nedelec element on a pyramid
+class Nedelec2PyrFiniteElement : public VectorFiniteElement
```
Should we keep this in the PR? It's a very element and the test has been commented out: https://github.com/mfem/mfem/pull/4460/files#diff-810b21fbfced27e1db37060dd1f22d060a494cfb7e6695aba41f95fd518bd357R178

mlstowell replied:
A fair question. This was added by request to aid in testing before the arbitrary order version was available. We probably no longer need it. If no one expresses a desire to keep it, I'll remove it before merging.

mlstowell commented:
@hughcars, I'd be very interested to hear any feedback you might have regarding the use of these new pyramid basis functions in your applications.

v-dobrev commented on fem/fe/fe_pos.cpp:
```diff
@@ -1042,6 +1042,435 @@ void H1Pos_WedgeElement::CalcDShape(const IntegrationPoint &ip,
+   Vector m_shape(nterms);
+#endif
+
+   CalcShape(order, ip.x, ip.y, ip.z, m_shape_1d.GetData(), m_shape.GetData());
```
This method seems to evaluate a lot more terms than necessary which raises the complexity with respect to the order. Have you considered using the terms in the following expression (these terms define the basis):

$$
1 = \sum_{k=0}^p \sum_{i=0}^{p-k} \sum_{j=0}^{p-k} \left[
        {p \choose k} {p-k \choose i} {p-k \choose j}
        \frac{z^k}{(1-z)^{p-k}}
        x^i (1-x-z)^{p-k-i}
        y^j (1-y-z)^{p-k-j} \right]
$$

Basically for every $k=0,...,p$, there are $(p-k+1)^2$ terms (the ones involving $x$ and $y$) to be evaluated and one can use the product structure to do that.

Or is the basis implemented here different? It has the same number of basis functions, so I suspect it is the same.

mlstowell replied:
This does look more efficient. I'll try to make use of it. Thanks!

mlstowell replied:
It looks like we don't have any examples or unit tests that use the H1_Pos bases. There seems to be one unit test, `fem/test_dgmassinv.cpp`, which exercises the first order positive L2 basis.

I propose we delay this change and put it in another PR with at least a unit test and perhaps an example.  What do you think (@v-dobrev, @tzanio, @psocratis, @hughcars)?

v-dobrev replied:
I don't insist on optimizing this now.

v-dobrev commented on fem/eltrans.cpp:
```diff
@@ -458,6 +460,8 @@ int IsoparametricTransformation::OrderW() const
       case FunctionSpace::Qk:
          return (FElem->GetOrder() * FElem->GetDim() - 1);
+      case FunctionSpace::Uk:
+         return (FElem->GetOrder() * FElem->GetDim());
```
I think this can actually return the same value as `Qk`.

mlstowell replied:
I had tried that initially but some integrals were causing trouble. I had assumed this was due to the different ways the Jacobian enters some of the integrals involving H(Curl) and H(Div) bases. I just reran the unit tests with the lower `OrderW` and they all pass. Those issues must have been resolved by a bug fix.

I'll lower this as you suggest. Thanks!

v-dobrev commented on fem/intrules.cpp:
```diff
@@ -1791,10 +1791,10 @@ IntegrationRule *IntegrationRules::TetrahedronIntegrationRule(int Order)
+   // rule for a cube which seems to be adequate for now. We should continue
+   // to search for a more appropriate integration rule designed specifically
+   // for pyramid elements.
+   const IntegrationRule &irc = Get(Geometry::CUBE, Order + 2);
```
This increase in the order is actually not necessary, if the definitions of
* `IsoparametricTransformation::OrderW()`,
* `IsoparametricTransformation::OrderJ()`, and
* `IsoparametricTransformation::OrderGrad()`,

for `Uk` use the same values as `Qk`. One can treat integration on the reference pyramid with `Uk` spaces as integration on the reference cube with `Qk` spaces -- in this case the term `(1-z)^2` becomes part of the determinant of the Jacobian of the mapping from the reference cube (through the Duffy map to the reference pyramid) to the physical pyramid. The reason is that under the Duffy map `Uk` on the pyramid maps to a subset of `Qk` on the reference element.

mlstowell replied:
@v-dobrev, thanks again for your advice on these integration rules. I was able to use the lower integration order as you suggest here. I did need to bump up the order by 2 in both the CurlCurl and DivDiv integrators when these `Uk` function spaces are present. No other changes were necessary to get both the unit tests and the convergence tests to pass.

v-dobrev commented on makefile:
```diff
@@ -377,6 +377,7 @@ MFEM_EXT_LIBS  ?= $(ALL_LIBS) $(LDFLAGS) $(LDFLAGS_INTERNAL)
 MFEM_LIBS      ?= $(if $(shared),$(BUILD_RPATH)) -L@MFEM_LIB_DIR@ -lmfem\
    @MFEM_EXT_LIBS@
 MFEM_LIB_FILE  ?= @MFEM_LIB_DIR@/libmfem.$(if $(shared),$(SO_VER),a)
+MFEM_LIBE_FILE ?= @MFEM_LIB_DIR@/libmfem-extras.$(if $(shared),$(SO_VER),a)
```
A new library file? Why is the code in it not put in `libmfem.*`? This is a big change that needs to be discussed if it is really needed.

mlstowell replied:
Sorry. I believe this is an artifact from a very old change which was reverted in favor of `miniapps/common`. I'll clean it up.

v-dobrev commented on mesh/mesh.cpp:
```diff
@@ -3622,9 +3622,9 @@ void Mesh::Make3D(int nx, int ny, int nz, Element::Type type,
-         ind[3] = VTX(x  , y+1, z  );
-         ind[4] = VTX(x  , y  , z+1);
-         ind[5] = VTX(x+1, y  , z+1);
+         ind[3] = VTX(x, y+1, z  );
+         ind[4] = VTX(x, y, z+1);
+         ind[5] = VTX(x+1, y, z+1);
```
Revert such unnecessary changes. Here and in other places.

psocratis commented on tests/unit/fem/test_fe_compatibility.cpp:
```diff
@@ -0,0 +1,1302 @@
+   {
+      dim = 3;
+      sdim = 3;
+      bdr_type = Geometry::TRIANGLE;
```
I don't get why the `bdr_type` is always a triangle. How about cubes, wedges, and pyramids?

mlstowell replied:
That argument is actually not used because I'm not creating any boundary elements in these meshes. I should probably leave it as ` Geometry::INVALID` to avoid confusion. I'll double check that this would work before making the change.

mlstowell replied:
It works fine a `Geometry::INVALID`. Thanks for catching this!

psocratis reviewed (commented):
This is an enormous amount of work! Thanks @mlstowell. 
I did look at all of it, but I'll need to spend a bit more time before I approve it. For now, I can confirm that all the unit tests (the old ones and the newly added ones) pass. I also tried ex1p and ex3p with visualization and they look okay. I suggest adding a couple of sample runs here and there. I also noticed that `UpdateOperator` is not yet supported. Any plans to add it in this PR?

hughcars commented:
I am trying to build with `MFEM_THREAD_SAFE` and `H1_BergotPyramidElement::CalcDShape` fails due to missing variables, it looks like they just need to be defined in the body. Similar missing variables in `fe_l2.cpp` in a few method bodies.

tzanio commented:
@brendankeith, a friendly reminder to please take another look!

brendankeith commented:
> @brendankeith, a friendly reminder to please take another look!

Thanks, @Tzanio. I am going to start again on Thursday!

hughcars commented:
More testing: When reading a higher-order pyramid mesh using this patch, I get a minimum side length of 0 (and thus inf kappa).
```
Parallel Mesh Stats:
...
            minimum     maximum
 h                0    0.109954
 kappa      1.03932         inf
 ```
 but after removing curvature 
 ```
             minimum     maximum
 h      0.000205315    0.109954
 kappa      1.03932     758.573
 ```
 This possibly due to the elements being too anisotropic/curved (removing the curvature here means interpolating Nodes onto `p=1` effectively), as the non-curved anisotropy is still relatively large. Will try and regenerate with less curvature any try again.

v-dobrev commented on mesh/mesh.cpp:
```diff
@@ -6675,7 +6676,8 @@ int Mesh::CheckElementOrientation(bool fix_it)
                   wo++;
                   if (fix_it)
                   {
-                     // how?
+                     mfem::Swap(vi[1], vi[3]);
+                     fo++;
```
For this to work we may need to change the `Finalize*` methods -- right now, it is assumed that only some mesh types (e.g. tet meshes) can be affected by this method, see https://github.com/mfem/mfem/blob/31a60b4476317ffc1b39421886fda798b2345de8/mesh/mesh.cpp#L3484-L3487

Similar change is implemented in PR #4607 for hexes and prisms, see the changes there in `Mesh::DoNodeReorder`, `Mesh::Finalize`, `Mesh::CheckElementOrientation`. In particular, for meshes with nodes, one probably needs to implement method `Mesh::GetPyramidOrientation` for the identity permutation and for the permutation that swaps vertices 1 and 3; this orientations then need to be supported in the H1 collection, see e.g. the changes in `H1_FECollection::DofOrderForOrientation` in PR #4607.

brendankeith commented on fem/fe/fe_h1.hpp:
```diff
@@ -148,6 +149,58 @@ class H1_WedgeElement : public NodalFiniteElement
                    DenseMatrix &dshape) const override;
 };
 
+class H1_FuentesPyramidElement
```
Can you add a reference to the paper as a comment (here or somewhere else)? I doubt people will recognize the elements by this name

mlstowell replied:
Yes, of course. I did mean to do this. Thanks for catching it.

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+namespace mfem
+{
+
+class FuentesPyramid
```
This is probably the file where the citation should go

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+   void phi_T(int p, Vector nu, const DenseMatrix &grad_nu,
+              DenseMatrix &u, DenseTensor &grad_u) const;
+   /*
+   static void E_E(int p, real_t s0, real_t s1, real_t *u,
```
Unused code?

brendankeith commented on fem/fe/fe_pyramid.cpp:
```diff
@@ -0,0 +1,1584 @@
+
+Vector FuentesPyramid::grad_lam1(real_t x, real_t y, real_t z)
+{
+   return Vector({CheckZ(z) ? - (1.0 - y - z) / (1.0 - z) : -0.5,
```
Where do the values at z = 1.0 come from?

If we take y = \alpha (1.0 - z), then we get \lim_{z\to \infty}  -(1.0 - y - z) / (1.0 - z) =  \alpha - 1.0 

I suppose you are setting \alpha = 0.5?

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+   static DenseMatrix grad_lam35(real_t x, real_t y, real_t z);
+   static DenseMatrix grad_lam45(real_t x, real_t y, real_t z);
+
+   static Vector lam15_grad_lam15(real_t x, real_t y, real_t z);
```
Suggested comment:
`lambda_ij = \lambda_i \nabla \lambda_j -  \lambda_j \nabla \lambda_i`

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+   static Vector lam35_grad_lam35(real_t x, real_t y, real_t z);
+   static Vector lam45_grad_lam45(real_t x, real_t y, real_t z);
+
+   /// Three component vectors associated with triangular faces
```
Are these explicit formulae for `s_0 \nabla s_1 \times \nabla s_2 + s_1 \nabla s_2 \times \nabla s_0 + s_2 \nabla s_0 \times \nabla s_1` ?

I tried to verify this guess, but there are so many terms that I think I would need to use Mathematica.

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+   { return Vector({lam1(x, y, z), lam4(x, y, z), lam5(x, y, z)}); }
+
+   /// Vector functions related to the normals to the triangular faces
+   static Vector lam125_grad_lam125(real_t x, real_t y, real_t z);
```
Are these explicit formulae for `s_0 \nabla s_1 \times \nabla s_2 + s_1 \nabla s_2 \times \nabla s_0 + s_2 \nabla s_0 \times \nabla s_1`?

I tried to verify this guess, but I think that I would need Mathematica.

mlstowell replied:
I have added a more detailed comment

brendankeith commented on fem/fe/fe_pyramid.cpp:
```diff
@@ -0,0 +1,1584 @@
+   MFEM_ASSERT(p >= 0, "Polynomial order must be zero or larger");
+   if (t > 0.0)
+   {
+      Poly_1D::CalcLegendre(p, x / t, u);
```
This could be numerically unstable when t is near 0

There is a recursive formula for these polynomials in the "Polynomials" table (Appendix E.1).

brendankeith commented on fem/fe/fe_pyramid.cpp:
```diff
@@ -0,0 +1,1584 @@
+   {
+      Poly_1D::CalcLegendre(p, x / t, u, dudx);
+      dudx[0] = 0.0;
+      dudt[0] = - dudx[0] * x / t;
```
`dudx[0] = 0` by the line above, so you can replace this with `dudt[0] = 0.0`

brendankeith commented on fem/fe/fe_pyramid.hpp:
```diff
@@ -0,0 +1,489 @@
+
+   static void CalcScaledLegendre(int p, real_t x, real_t t,
+                                  Vector &u);
+   static void CalcScaledLegendre(int p, real_t x, real_t t,
```
It looks like this function is only used to compute the derivatives of the integrated Legendre functions.

Computing the derivatives of the integrated Legendre functions as described in https://github.com/mfem/mfem/pull/4460#discussion_r1976701476 is more computationally efficient and will allow this function to be removed

brendankeith commented on fem/fe_coll.cpp:
```diff
@@ -1874,21 +1873,42 @@ H1_FECollection::H1_FECollection(const int p, const int dim, const int btype)
+         else
+         {
+            H1_dof[Geometry::PYRAMID] = (p-1)*(p-2)/2;
+            // H1_dof[Geometry::PYRAMID] = 0;
```
Remove comment

brendankeith commented on fem/fe_coll.cpp:
```diff
@@ -1874,21 +1873,42 @@ H1_FECollection::H1_FECollection(const int p, const int dim, const int btype)
+            }
          }
-         H1_Elements[Geometry::PYRAMID] = new LinearPyramidFiniteElement;
+         // H1_Elements[Geometry::PYRAMID] = new LinearPyramidFiniteElement;
```
Unecessary comment

brendankeith commented on fem/fe_coll.cpp:
```diff
@@ -2230,20 +2240,33 @@ L2_FECollection::L2_FECollection(const int p, const int dim, const int btype,
+         }
       }
-      L2_Elements[Geometry::PYRAMID] = new P0PyrFiniteElement;
+      // L2_Elements[Geometry::PYRAMID] = new P0PyrFiniteElement;
```
Remove comment

brendankeith commented on fem/fe_coll.cpp:
```diff
@@ -2475,8 +2490,9 @@ RT_FECollection::RT_FECollection(const int order, const int dim,
 
-      RT_Elements[Geometry::PYRAMID] = new RT0PyrFiniteElement(false);
-      RT_dof[Geometry::PYRAMID] = 0;
+      // RT_Elements[Geometry::PYRAMID] = new RT0PyrFiniteElement(false);
```
Remove

brendankeith commented on fem/fe_coll.cpp:
```diff
@@ -2901,25 +2907,17 @@ ND_FECollection::ND_FECollection(const int p, const int dim,
 
-      ND_Elements[Geometry::PYRAMID] = new Nedelec1PyrFiniteElement;
-      ND_dof[Geometry::PYRAMID] = 0;
+      // ND_Elements[Geometry::PYRAMID] = new Nedelec1PyrFiniteElement;
```
Delete comments

brendankeith commented on fem/fe/fe_pos.hpp:
```diff
@@ -256,6 +257,62 @@ class H1Pos_WedgeElement : public PositiveFiniteElement
 
 
+/// Arbitrary order H1 elements in 3D utilizing the Bernstein basis on a pyramid
+class H1Pos_PyramidElement : public PositiveFiniteElement, FuentesPyramid
```
Can you include a reference for this element?

mlstowell replied:
I din't have a reference for this but I'll expand the comment.

brendankeith commented on fem/fe/fe_nd.cpp:
```diff
@@ -1581,6 +1581,954 @@ void ND_WedgeElement::CalcCurlShape(const IntegrationPoint &ip,
+               }
+
+      // Family IV
+      // Re-using mu from Family I
```
I think you are reusing mu from Family II

mlstowell replied:
Good catch!

brendankeith commented on fem/fe/fe_rt.cpp:
```diff
@@ -1264,6 +1264,743 @@ void RT_WedgeElement::CalcDivShape(const IntegrationPoint &ip,
+      E_Q(p, mu01(z, xy, 2), grad_mu01(z, xy, 2),
+          mu01(z, xy, 1), grad_mu01(z, xy, 1), E_ijk, dE_ijk);
+      // Re-using phi_E from Family I
+      // phi_E(p, mu01(z), grad_mu01(z), phi_k, dphi_k);
```
Delete commented command

mlstowell commented:
> This is a great contribution, that many people will benefit from. Thank you, @mlstowell, and sorry that it took me so long.
> 
> I think most of the PR is in order, but some of the functions in `fe_pyramid.cpp` could be improved. I reviewed all the primary files but skipped a few unit tests. Some of the tests I reviewed had commented code, so I suggest each of them to ensure that any commented code I missed is removed.
> 
> I add numerous minor comments, but will happily approve the PR once they are addressed

Thank you, @brendankeith, for all of your very useful comments and suggestions. I believe I have addressed all of the comments related to documentation improvements and removal of dead code.

Later today I will create a new branch to address the improvements to the polynomials that you suggest. This is mainly so that I can more easily compare the new implementations to the current ones. However, it may also give us the chance to merge this `pyramid-dev` branch into MFEM in time for our next release without waiting for the new implementations to be fully tested.

mlstowell commented on fem/intrules.cpp:
```diff
@@ -1791,23 +1791,39 @@ IntegrationRule *IntegrationRules::TetrahedronIntegrationRule(int Order)
+   PyramidIntRules[Order]->SetOrder(Order);
 
-   for (int k=0; k<npts; k++)
+   if (npts == 1)
```
Good point. I should have checked that...

mlstowell commented:
@brendankeith, would you approve of postponing the polynomial improvements you suggest until a follow-up PR? This is just in the interest of preparing this PR for the next release ASAP.

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Re-merged in `next` for testing...

tzanio reviewed (approved):
Thank you for all your hard work on this awesome addition, @mlstowell! 🚀

hughcars commented:
> I am trying to build with `MFEM_THREAD_SAFE` and `H1_BergotPyramidElement::CalcDShape` fails due to missing variables, it looks like they just need to be defined in the body. Similar missing variables in `fe_l2.cpp` in a few method bodies.
> 
> ```diff
> diff --git a/fem/fe/fe_h1.cpp b/fem/fe/fe_h1.cpp
> index 14136ad9f..93d45ad91 100644
> --- a/fem/fe/fe_h1.cpp
> +++ b/fem/fe/fe_h1.cpp
> @@ -1821,6 +1821,8 @@ void H1_BergotPyramidElement::CalcDShape(const IntegrationPoint &ip,
>     const int p = order;
>  
>  #ifdef MFEM_THREAD_SAFE
> +   Vector shape_x(p + 1), shape_y(p + 1), shape_z(p + 1);
> +   Vector dshape_x(p + 1), dshape_y(p + 1), dshape_z(p + 1), dshape_z_dt(p + 1), u(p + 1);
>     DenseMatrix du(dof, dim);
>  #endif
>     real_t x = (ip.z < 1.0) ? (ip.x / (1.0 - ip.z)) : 0.0;
> diff --git a/fem/fe/fe_l2.cpp b/fem/fe/fe_l2.cpp
> index 2ba4c0b21..8fe51340c 100644
> --- a/fem/fe/fe_l2.cpp
> +++ b/fem/fe/fe_l2.cpp
> @@ -1199,6 +1199,7 @@ void L2_BergotPyramidElement::CalcDShape(const IntegrationPoint &ip,
>     Vector dshape_x(p + 1);
>     Vector dshape_y(p + 1);
>     Vector dshape_z(p + 1);
> +   Vector dshape_z_dt(p + 1);
>     DenseMatrix du(dof, dim);
>  #endif
> ```
> 
> worked but could do with checking. Without this fix I can't build this branch.

Fixed in https://github.com/mfem/mfem/pull/4780#pullrequestreview-2743608646
